bionic/executor.py

```diff
@@ -6,19 +6,83 @@
 """
 
 import copy
+from functools import partial
 import logging
+from multiprocessing.managers import SyncManager
 import queue
 import sys
 import threading
 import traceback
+from uuid import uuid4
 
-from multiprocessing.managers import SyncManager
-
+from .aip.task import Task
 from .deps.optdep import import_optional_dependency
 from .utils.misc import oneline, SynchronizedSet
 
 
-class Executor:
+class _BaseExecutor:
+    def submit(self, fn, *args, **kwargs):
+        """
+        Submits the function to the executor and returns a Future object.
+        """
+        pass
+
+    def create_synchronized_set(self):
+        """
+        Creates a set that is used by TaskKeyLogger for progress reporting.
+        """
+        pass
+
+    def start_logging(self):
+        """
+        Sets up anything related to logging in the current process for logs to
+        work for the submitted tasks.
+        """
+        pass
+
+    def stop_logging(self):
+        """
+        Stops any logging work in the current process that is not needed anymore.
+        """
+        pass
+
+    def set_resource(self, resource):
```

Comment:
Hmmm.... it's not clear to me why this base class is better than just having different APIs for the different executor types. If AipExecutor requires a resource to run something remotely, why not just make that part of the API?

Reply:
I thought this abstraction made the caller code easier, but that is only when we use only one kind of executor, which means either parallel or aip (not both). But if we support both parallel and aip execution on the same flow, then this abstraction doesn't help in any way. Then it would be easier to do what you suggested, have different APIs for different executor types.